### PR TITLE
Allow fuzzer to persist repro data on disk upon failure

### PR DESCRIPTION
**Pull request by tanjialiang**
### Summary
Adds functionalities to:
* allow fuzzer to persist problematic repro data (both input and result vector) 
  and expression on disk. 
* allow fuzzer to read from on-disk repro data and expression to reproduce 
  error.

### Tests:

Initial run to get failed run, persisting on disk:
`/Users/jtan6/git-repos/presto/presto-native-execution/cmake-build-debug/velox/velox/expression/tests/velox_expression_fuzzer_test --repro_persist_path /tmp/abc --seed 1813425897 --v=1 --logtostderr=1`

Run log:
https://gist.github.com/tanjialiang/c97820551da3dd14a9a73625a1b5615a

Second run with first run's data:
`/Users/jtan6/git-repos/presto/presto-native-execution/cmake-build-debug/velox/velox/expression/tests/velox_expression_fuzzer_test --repro_data_restore_path /tmp/abc/velox_vector_Fh7WYR --repro_sql_restore_path /tmp/abc/velox_sql_ZjSKgw --steps 1 --v=1 --logtostderr=1`
 
Run log:
https://gist.github.com/tanjialiang/f4b4d70a416d1ff0ca280309496cf878

### CODE REVIEW

File: velox/expression/tests/ExpressionFuzzer.cpp [hunk 1]
```diff
@@ -638,15 +680,40 @@ class ExpressionFuzzer {
...
+      if (!FLAGS_repro_persist_path.empty()) {
+        std::string inputPath;
+        std::string resultPath;
+        std::string sqlPath;
+        exec::ExprSaver::saveData(
+            rowVector.get(),
+            FLAGS_repro_persist_path.c_str(),
+            inputPath,
+            inputPath);
+        if (copiedResult) {
```
Comment (mbasmanova) [inline velox/expression/tests/ExpressionFuzzer.cpp:704]: Why do we need to save the result vector? BTW, copying flattens the vector and we lose information about the encodings.

File: velox/expression/tests/ExpressionFuzzer.cpp [hunk 2]
```diff
@@ -665,6 +732,39 @@ class ExpressionFuzzer {
 
  public:
   void go() {
+    // Go to on-disk repro mode.
```
Comment (mbasmanova) [inline velox/expression/tests/ExpressionFuzzer.cpp:735]: Do we need to couple the repro logic with the Fuzzer? Would it be more flexible to create a separate CLI that takes a data and SQL paths and executes the expression o the data?

File: velox/expression/Expr.h [hunk 3]
```diff
@@ -63,25 +63,64 @@ struct ExprStats {
...
+  /// Returns true if persistence is successful, false otherwise. In the case of
+  /// unsuccessful persistence, the reason will be returned by failMsg.
+  static bool saveSql(
+      const std::string& sql,
+      const char* FOLLY_NONNULL basePath,
+      std::string& storedPath,
+      std::string& failMsg);
+};
+
+class ExprRestorer {
```
Comment (mbasmanova) [inline velox/expression/Expr.h:89]: Do we need to couple the logic of reading vector and SQL from files with expression evaluation?

File: velox/expression/Expr.h [hunk 4]
```diff
@@ -63,25 +63,64 @@ struct ExprStats {
   }
 };
 
+class ExprSaver {
+ public:
+  /// Persist the passed in 'RowVector' on disk. The data will be persisted at
+  /// $basePath/velox_vector_xxx that will be returned by storedPath.
+  /// Returns true if persistence is successful, false otherwise. In the case of
+  /// unsuccessful persistence, the reason will be returned by failMsg.
+  static bool saveData(
```
Comment (mbasmanova) [inline velox/expression/Expr.h:72]: Do we need to add more functionality to expression evaluation engine. Can this logic be added to the Fuzzer directly?

**Review by mbasmanova — COMMENTED** [review]
@tanjialiang Some initial questions.

File: velox/expression/tests/ExpressionFuzzer.cpp (hunk 1 above)
Comment (tanjialiang) [reply velox/expression/tests/ExpressionFuzzer.cpp:704]: Dirty initialized result vector sometimes can be the key to some failures, just like last time I fixed the bug greatest() has by not clearing result vector before writing into it.
Are there any deep copy that preserves everything?

File: velox/expression/tests/ExpressionFuzzer.cpp (hunk 2 above)
Comment (tanjialiang) [reply velox/expression/tests/ExpressionFuzzer.cpp:735]: I think it is better to keep them together because we need to see the discrepancy of common path vs simple path when reproducing. For example when a fuzzer run fail due to result not matching, then we need to compare the results  between common and simple path to debug.

File: velox/expression/Expr.h (hunk 3 above)
Comment (tanjialiang) [reply velox/expression/Expr.h:89]: I could not find a better place to put these saver and restorer. They are too small to be put in a separate file..

File: velox/expression/Expr.h (hunk 4 above)
Comment (tanjialiang) [reply velox/expression/Expr.h:72]: ExprExpcetionContext needs this logic as well, which is in Expr.h

File: velox/expression/tests/ExpressionFuzzer.cpp (hunk 1 above)
Comment (mbasmanova) [reply velox/expression/tests/ExpressionFuzzer.cpp:704]: > Are there any deep copy that preserves everything?

Only serializing using the new VectorSaver.

File: velox/expression/Expr.h (hunk 3 above)
Comment (mbasmanova) [reply velox/expression/Expr.h:89]: Can they go into VectorSaver?

File: velox/expression/tests/ExpressionFuzzer.cpp (hunk 2 above)
Comment (mbasmanova) [reply velox/expression/tests/ExpressionFuzzer.cpp:735]: Maybe we can have this functionality available separately as well. For example, we could add a flag to tell whether to use common or simplified expression evaluation path.

One concern about putting everything into the Fuzzer is it makes the Fuzzer very large ands dependencies such as DuckDB.

File: velox/expression/tests/ExpressionFuzzerTest.cpp [hunk 5]
```diff
@@ -51,8 +51,7 @@ int main(int argc, char** argv) {
       // (since HLL is a user defined type), and end up trying to use
       // cardinality passing a VARBINARY (since HLL's implementation uses an
       // alias to VARBINARY).
-      "cardinality",
-      "neq"};
```
Comment (mbasmanova) [inline velox/expression/tests/ExpressionFuzzerTest.cpp:55]: Unrelated change?

File: velox/expression/Expr.cpp [hunk 6]
```diff
@@ -375,46 +376,117 @@ std::optional<std::string> generateFilePath(
...
   } catch (...) {
-    sqlPath_ =
+    failMsg =
         fmt::format("Failed to save expression SQL to {}.", sqlPath.value());
+    return false;
+  }
+  return true;
+}
+
+void ExprExceptionContext::persistDataAndSql(const char* basePath) {
```
Comment (mbasmanova) [inline velox/expression/Expr.cpp:443]: Did you end up not using ExprExceptionContext in the Fuzzer? Should we move it back into the .cpp file?

File: velox/expression/tests/ExpressionFuzzer.cpp [hunk 7]
```diff
@@ -580,30 +614,39 @@ class ExpressionFuzzer {
...
     }
 
+    // Store data and expression in case of reproduction.
+    std::shared_ptr<BaseVector> copiedResult;
+    std::string sql;
+    // Deep copy to preserve the initial state of result vector.
+    if (!FLAGS_repro_persist_path.empty()) {
+      copiedResult =
+          resultVector ? BaseVector::copy(*resultVector) : resultVector;
+      sql = plan->toString();
```
Comment (mbasmanova) [inline velox/expression/tests/ExpressionFuzzer.cpp:624]: Shouldn't this call plan->toSql() ?

File: velox/expression/Expr.h (hunk 3 above)
Comment (tanjialiang) [reply velox/expression/Expr.h:89]: It deals with expressions as well which is technically not vector. The reason I put it here is kind of natural because in ExprExceptionContext it has both data and sql. Maybe we can put all ExprExceptionContext, the saver and restorer in a separate file, maybe named ExprUtil.h?

File: velox/expression/tests/ExpressionFuzzerTest.cpp (hunk 5 above)
Comment (tanjialiang) [reply velox/expression/tests/ExpressionFuzzerTest.cpp:55]: Yeah, thanks for catching it.

File: velox/expression/Expr.cpp (hunk 6 above)
Comment (tanjialiang) [reply velox/expression/Expr.cpp:443]: Right I separated the persistence logic out. Any reason we want to move it back?

File: velox/expression/tests/ExpressionFuzzer.cpp (hunk 7 above)
Comment (tanjialiang) [reply velox/expression/tests/ExpressionFuzzer.cpp:624]: I think for ITypedExpr there is no toSql() call. The equivalent is just toString() if I'm not wrong.

File: velox/expression/tests/ExpressionFuzzer.cpp (hunk 2 above)
Comment (tanjialiang) [reply velox/expression/tests/ExpressionFuzzer.cpp:735]: Thanks, that makes sense. Quick question, what is the downside of making Fuzzer a little larger?

File: velox/expression/tests/ExpressionFuzzer.cpp (hunk 7 above)
Comment (tanjialiang) [reply velox/expression/tests/ExpressionFuzzer.cpp:624]: This was what I did before. I think somehow after compilation or execution the expressions got evaluated already and is no long the original structure.

Comment (tanjialiang) [reply velox/expression/tests/ExpressionFuzzer.cpp:624]: Let me see what I can do here.

Comment (tanjialiang) [reply velox/expression/tests/ExpressionFuzzer.cpp:624]: Ah. The reason it does not preserve original expression when using exprSet is because it enables constant folding. Disabling it makes it work. I will make the changes.

**Comment by tanjialiang** [comment]
1. extract file creation logic out, common/file/..
2. save expr could be generic as stirng opertion. remove expr saver api to let caller to directly use VectorSavor
3. extract the result/exception comparison part of fuzzer out
4. use 3. to create another executable to specifically reproduce from saved files.
5. [next item] add support for crash persistency (write to file every iteration, reuse ostream to allow all in-memory operation)

**Review by kagamiori — COMMENTED** [review]
Hi Jialiang, Thank you for helping add this feature. I left a few questions.

File: velox/expression/Expr.cpp [hunk 8]
```diff
@@ -375,46 +376,147 @@ std::optional<std::string> generateFilePath(
...
+    failMsg = fmt::format(
+        "Failed to save input vector to {}. Reasons: '{}'",
+        dataPath.value(),
+        e.what());
+    return false;
+  } catch (...) {
+    failMsg =
+        fmt::format("Failed to save input vector to {}.", dataPath.value());
+    return false;
+  }
```
[the remark this answers is not in the document]
Comment (tanjialiang) [reply velox/expression/Expr.cpp:437]: I'm not sure. Catching it just in case.

[the remark this answers is not in the document]
Comment (tanjialiang) [reply velox/expression/Expr.cpp:437]: I'm not sure. Catching it just in case.

File: velox/expression/tests/ExpressionFuzzer.cpp [hunk 9]
```diff
@@ -580,30 +615,46 @@ class ExpressionFuzzer {
...
     SelectivityVector rows{rowVector ? rowVector->size() : 1};
 
-    // Randomize initial result vector data to test for correct null and data
-    // setting in functions.
-    if (vectorFuzzer_.coinToss(0.5)) {
-      commonEvalResult[0] = vectorFuzzer_.fuzzFlat(plan->type());
-    }
-
     // Execute with common expression eval path.
+    std::shared_ptr<exec::ExprSet> exprSetCommon;
```
[the remark this answers is not in the document]
Comment (tanjialiang) [reply velox/expression/tests/ExpressionFuzzer.cpp:648]: We could do that yeah.

File: velox/expression/tests/ExpressionFuzzer.cpp [hunk 10]
```diff
@@ -638,15 +688,40 @@ class ExpressionFuzzer {
...
+    } catch (...) {
+      if (!FLAGS_repro_persist_path.empty()) {
+        std::string inputPath;
+        std::string resultPath;
+        std::string sqlPath;
+        exec::ExprSaver::saveData(
+            rowVector.get(),
+            FLAGS_repro_persist_path.c_str(),
+            inputPath,
+            inputPath);
```
[the remark this answers is not in the document]
Comment (tanjialiang) [reply velox/expression/tests/ExpressionFuzzer.cpp:711]: They were a single argument. I deliberately separated them for clarity. Also @kagamiori this PR is going to be abandoned and a new series is been developed in a way described as above bullet points show. Can you also help to take a look? https://github.com/facebookincubator/velox/pull/2741
Thanks!
